### ISSUE TITLE
Variables weren't shown

### ISSUE THREAD

**Issue by kopaygorodsky** [derekparker/delve#738]
`Delve Debugger 0.12.1`
`GOlang 1.7.5 darwin/amd64`
`MacOS  10.12.3`

I'm trying to debug my application,  it's simple KMP algorithm, and I can't see any local variables except arguments of the function. 
Here is an issue with the full explanation was created for VSCode plugin, I think it should be referenced. 
https://github.com/Microsoft/vscode-go/issues/793

**Comment by aarzilli** [comment]
Failed to get threads doesn't look like an error that delve prints, without knowing what we returned to VSCode it's hard to know what happened here and who's at fault.

**Comment by kopaygorodsky** [comment]
@aarzilli, I agree that it's not a big problem, but a bug with variable exists.

**Comment by SteelPhase** [comment]
The below is pertaining to Failed to eval expression from vscode-go.

The error text delve is returning over RPC to vscode-go is `could not find symbol value for debug`. 
this is coming back from EvalSymbol RPC call. 

the args provided by vscode are 
```
{
 "symbol": "debug",
 "scope": {
  "goroutineID": 37,
  "frame": 0
 }
}
```

If you want more details let me know, and I can try and get them for you, but this does look like all that's being provided for the RPC call that's failing.

**Comment by SteelPhase** [comment]
In my testing it does not affect all local variables. And it does not generate a specific error.
VariablesRequest just seems to not return the missing variable, and EvalSymbol returns `could not find symbol value for %s~ %s being the name of the variable you're trying to Eval.

If i find some more time to test this, i'll see if i can generate a test case that can replicate this on it's own.

**Comment by aarzilli** [comment]
If the program was compiled with optimizations that could be the reason.

**Comment by SteelPhase** [comment]
The debug executable used with delve was generated at the time of debugging within vscode. I've never looked at how they are generated, but i will get back to you on what I find.

**Comment by SteelPhase** [comment]
vscode uses delve debug to execute and debug the code. So i'd hope delve debug would generate an assembly it could debug correctly.

**Comment by aarzilli** [comment]
@SteelPhase your bug seems very different from what @kopaygorodsky reported. What operating system are you using? Version of go? What does the function that lacks a variable looks like? Is it in the main package or on another package? Do you have a simple way to reproduce it?

**Comment by derekparker** [comment]
> So i'd hope delve debug would generate an assembly it could debug correctly.

The Go compilation tool chain produces the actual binary, not Delve. Delve does however use various compiler flags by default to produce a binary that is optimal for debugging.

Do you think either of you @SteelPhase or @kopaygorodsky could provide a gist of some code that causes Delve to exhibit this behavior, or possibly provide a failing test case so that we can try and dig into this further and see if it is a bug within Delve or with the debug information the compilation tool chain emits.

**Comment by SteelPhase** [comment]
I've been trying to replicate this in a small snippet but I have yet to figure out what pieces need to be in place to cause it. @aarzilli I am getting the same situation as @kopaygorodsky is that he documented here Microsoft/vscode-go#793. Just wanted to help provide as much detail as I can to help.

**Comment by kopaygorodsky** [comment]
@derekparker 
In this issue, I described the problem. Failing test is in the first screen shot. 
[https://github.com/Microsoft/vscode-go/issues/793](url) 
I've just copied the code to the gist. 
[https://gist.github.com/kopaygorodsky/264d3d6f0b4745b291951577dae18c01](url)

**Comment by kopaygorodsky** [comment]
hey!
I've just got another weird bug in debugger. 
On this screenshot you can see variable 'len', but there is no 'len' variable in the code. 
![image](https://cloud.githubusercontent.com/assets/2415637/23376207/026dd7a8-fd34-11e6-8433-10d56ac58bd7.png)

**Comment by aarzilli** [comment]
Your code is in a package other than main, the compiler is picking up the binary compiled for that package without -N -l.

**Comment by kopaygorodsky** [comment]
@aarzilli, so it's normal behavior.  
After I've restarted three times it works.
Thanks

**Comment by aarzilli** [comment]
No, but the fix needs to be in the compiler.

**Comment by kopaygorodsky** [comment]
@aarzilli 
"If the program was compiled with optimizations that could be the reason." -  I got your point. Thanks.

**Comment by kopaygorodsky** [comment]
just one notice: this bug disappears when I use fmt package, so if I don't see some variables, just add some fmt.Println and it will work.

**Comment by kopaygorodsky** [comment]
Wow, you are working on this issue, that's really nice. Thanks!
